scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamBackend.scala

```diff
@@ -115,6 +115,14 @@ object BeamPlanner {
             CoGroupedOp(cg, ops)
           }
           go(cg)
+        case (Fork(input), rec) =>
+          rec(input)
```

Comment:
btw: if you could link to any documentation here that says that beam can handle fanouts well or under what circumstances you need to annotate them in any way, that would be useful.

A fork node is added by the planner if a node has more than one downstream consumer.

Reply:
I assumed it would but let me dig into that.

Reply:
Just to be clear, is it ok to assume that the nodes leading up to this Fork will be processed only once by the backend and that their resulting PCollection instance will be used for all downstream nodes? So if there is re-computation it's entirely on Beam/Dataflow for re-computing a PCollection?

Reply:
Yes. That's what the memoization is for. The whole `rec` bit is about caching.

Reply:
Cool, just wanted to confirm. I'm observing the following:

- Handwritten Beam code running on Dataflow does reuse PCollections and does a fan-out to multiple outputs. Not sure there are rules or hints can be passed but seems to just work. Will try to find out if there are limitations or if always reuses.
- The Scalding to Beam conversion for a real job is not doing reuse and is recomputing everything
  - One of our sources implementation doesn't have equality implemented and we though this was messing up memoization, we think we fixed it but still no reuse.
  - Trying even simpler cases, with TypedPipe.from(List(...)) and still can't get any reuse.
  
I will continue experimenting and narrowing it down, trying to understand if there is an issue with the conversion or in some assumption about Beam/Dataflow.

Reply:
Ok something like this does the trick:

```scala
  final case class TransformBeamOp[A, B](
    source: BeamOp[A],
    f: PTransform[PCollection[A], PCollection[B]],
    kryoCoder: KryoCoder,
    name: String
  ) extends BeamOp[B] {

    private lazy val memo = Memoize.function[Pipeline, PCollection[B]] {
      case (pipeline, _) =>
        val pCollection: PCollection[A] = widenPCollection(source.run(pipeline))
        pCollection.apply(name, f).setCoder(kryoCoder)
    }

    def run(pipeline: Pipeline): PCollection[B] = memo(pipeline)
  }
  ```

@johnynek do you think this approach makes sense? There are only a handful of BeamOp's so it isn't too much work. We could do the memo in the parent trait but the type of the elements would be as precise. We could cast or use some abstract types if feeling fancy, or just duplicate the impl of the `run`.

Reply:
yeah, I would probably do something like:

```scala

protected def runNoCache(p: Pipeline): PCollection[_ <: A]
private val resultCache = scala.collection.mutable.Map.empty[Pipeline, PCollection[_ <: A]]
  
final def run(p: Pipeline): PCollection[_ <: A] =
  resultCache.getOrElseUpdate(p, runNoCache(p))
```

and then change all the current `def run` into `def runNoCache`

Reply:
This works by the way, got reuse on Dataflow.

Reply:
Okay, good. Sorry my message missed yours. I would still do the `def runNoCache` on the base trait (but you are right, using `Memoize.function` is safer actually we found some bugs trying to do memoization with mutable Map in scala, so internally it uses a var + immutable map).